nwb_conversion_tools/datainterfaces/icephys/abf/abfdatainterface.py

```diff
@@ -0,0 +1,142 @@
+            "recordings might be wrongly stored."
+        )
+        return neo_reader._axon_info["rec_datetime"]
+
+
+class AbfInterface(BaseIcephysInterface):
+    """ABF IcephysInterface based on Neo AxonIO."""
+
+    neo_class = AxonIO
+
+    @classmethod
+    def get_source_schema(cls):
+        source_schema = super().get_source_schema()
+        source_schema["properties"]["file_paths"] = dict(
+            type="array",
+            minItems=1,
+            items={"type": "string", "format": "file"},
+            description="Array of paths to ABF files.",
+        )
+        source_schema["properties"]["metadata_file_path"] = dict(
+            type="string", format="file", description="Path to JSON file containing metadata for this experiment."
+        )
+        return source_schema
+
+    def __init__(self, file_paths: list, metadata_file_path: str = None):
+        # TODO: add docstring for init
+        super().__init__(file_paths=file_paths)
+        self.source_data["metadata_file_path"] = metadata_file_path
+
+    def get_metadata(self):
+        metadata = super().get_metadata()
+
+        metafile_data = dict()
+        metafile = self.source_data["metadata_file_path"]
+        if metafile is not None and Path(metafile).is_file():
+            with open(metafile) as json_file:
+                metafile_data = json.load(json_file)
+
+        # Extract start_time info
+        first_reader = self.readers_list[0]
+        first_session_time = get_start_datetime(neo_reader=first_reader)
+        session_start_time = first_session_time.strftime("%Y-%m-%dT%H:%M:%S%z")
+
+        # NWBFile metadata
+        if "NWBFile" not in metadata:
+            metadata["NWBFile"] = dict()
+        metadata["NWBFile"].update(
+            session_start_time=session_start_time, experimenter=[metafile_data.get("experimenter", "")]
+        )
+
+        # Subject metadata
+        metadata["Subject"] = dict(
+            subject_id=metafile_data.get("subject_id", ""),
+            species=metafile_data.get("species", ""),
+            sex=metafile_data.get("sex", "U"),
+        )
+        if metafile_data.get("dob", None):
+            metadata["Subject"].update(date_of_birth=metafile_data.get("dob"))
+
+        # LabMetadata
+        metadata["ndx-dandi-icephys"] = dict(
+            # Required fields for DANDI
+            cell_id=metafile_data.get("cell_id", ""),
+            slice_id=metafile_data.get("slice_id", ""),
+            # Lab specific metadata
+            targeted_layer=metafile_data.get("targeted_layer", ""),
+            inferred_layer=metafile_data.get("estimate_laminate", ""),
+        )
+
+        # Recordings sessions metadata (one Session is one abf file / neo reader)
+        metafile_sessions = metafile_data.get("recording_sessions", dict())
+        metadata["Icephys"]["Sessions"] = list()
+
+        # Extract useful metadata from each reader in the sequence
+        i = 0
+        ii = 0
+        iii = 0
```

Comment:
Using explicit outer counters is definitely the 'easiest' approach, but if you're interested in the Pythonic code-reducing approach, you can wrap each of the iterables in the `for` loops below with `enumerate(..., start=xx)` where `xx` is the proper value depending on the outer loop counters to achieve this.